### PR TITLE
Add tensor_mechanics version of XFEMMaterialTensorMarkerUserObject

### PR DESCRIPTION
**Pull request by bwspenc**
ref #9604

### CODE REVIEW

**Comment by jiangwen84** [comment]
It looks good 👍

**Review by jiangwen84 — COMMENTED** [review]
It looks good 👍

File: modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    tensor_quantities.reserve(numqp);
+    Real max_quantity = 0;
+    std::vector<Point> directions(numqp);
+    for (unsigned int qp = 0; qp < numqp; ++qp)
+    {
+      tensor_quantities[qp] = RankTwoScalarTools::getQuantity(
+          _tensor[qp], _scalar_type, _point1, _point2, _q_point[qp], directions[qp]);
+      if (directions[qp](0) == 0 && directions[qp](1) == 0 && directions[qp](2) == 0)
+        mooseError("Direction has zero length in XFEMRankTwoTensorMarkerUserObject");
+      if (tensor_quantities[qp] > max_quantity)
```
Comment (permcody) [inline modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C]: Minor: Is it possible for _any_ user selected tensor_quantities value to be negative? My guess is no, but if it ever is, this branch won't be entered, which under a worst case scenario could cause you to access directions[numeric_limists<unsigned int>::max()] in the next branch below (if _threshold was also poorly chosen). You are clearly making at least one range assumption here, it probably should be asserted. Also, you might consider a range check on _threshold as well.

Comment (bwspenc) [inline modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C]: @permcody Good point.  I can think of cases where you might want the threshold to be negative, so I revised the code accordingly.

**Comment by bwspenc** [comment]
Oh yeah, documentation... I'll work on that!

**Comment by bwspenc** [comment]
@jasondhales I think this is ready to go (assuming the tests pass).  I addressed the issue pointed out by @permcody by enforcing that the threshold is always positive. The TM and SM versions of this test use a common gold file, and I added a few extra variants of the test for TM.

**Comment by bwspenc** [comment]
Now that I've resolved the test issues, I think this is good to go.

**Comment by bwspenc** [comment]
Does anyone know what's up with the 'Test Intel' tests being cancelled? @brianmoose @permcody? I tried invalidating it, but that didn't seem to do anything.

There was an issue with this commit earlier today -- a conflict because of the change to the test directories. That has been straightened out. Aside from the Intel tests not running, I think this is ready to go.

**Comment by permcody** [comment]
We've been having several problems with the testing infrastructure today. Not only have we run out of disk space on some of the clients (a somewhat regular occurrence), but we've also had a few networking and load issues with the CIVET server. We are looking into the later now. No need to do anything, we'll sort it all out. The TigerTeam activity has pushed the systems to their limits several times over the past 24 hours.

File: modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+bool
+XFEMRankTwoTensorMarkerUserObject::doesElementCrack(RealVectorValue & direction)
+{
+  bool does_it_crack = false;
+  unsigned int numqp = _qrule->n_points();
+
+  if (_average)
+  {
+    Real average_threshold;
```
Comment (permcody) [inline modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C]: You aren't initializing this Real! The first assignment is a `+=` which could give you garbage.

File: modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    unsigned int max_index = std::numeric_limits<unsigned int>::max();
+    Real max_ratio = 0.0;
+    std::vector<Point> directions(numqp);
+    for (unsigned int qp = 0; qp < numqp; ++qp)
+    {
+      if (_threshold[qp] <= 0.0)
+        mooseError("Threshold must be strictly positive in XFEMRankTwoTensorMarkerUserObject");
+      const Real tensor_quantity = RankTwoScalarTools::getQuantity(
+          _tensor[qp], _scalar_type, _point1, _point2, _q_point[qp], directions[qp]);
+      if (directions[qp](0) == 0.0 && directions[qp](1) == 0.0 && directions[qp](2) == 0.0)
```
Comment (permcody) [inline modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C]: Maybe use `directions[qp].size() == 0.0` or a fuzzy comparison?

**Review by permcody — CHANGES_REQUESTED** [review]
I spotted one potential bug. Can you take a look?

File: modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C (hunk 2 above)
Comment (bwspenc) [inline modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C]: Doh!  I'll fix that.

File: modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C (hunk 3 above)
Comment (bwspenc) [inline modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C]: Sure.  I'll do that.  These really will be set to zero in this case, but I'll make it fuzzy.

**Comment by bwspenc** [comment]
@permcody I addressed your requested changes.

**Review by permcody — APPROVED** [review]

**Pull request merged**